### PR TITLE
Fixed crashes on parsing and serialization

### PR DESCRIPTION
**Pull request by Sergeeeek**
It needs to be compiled with development version of zig, but diagnostics work now.

### CODE REVIEW

File: src/protocol.zig [hunk 1]
```diff
@@ -14,37 +14,47 @@ pub const LspError = error{
...
-        const line = stream.readUntilDelimiterOrEof(buffer[0..], '\n') catch |err| switch (err) {
+        const line = stream.readUntilDelimiterOrEof(buffer[0..], '\r') catch |err| switch (err) {
             error.StreamTooLong => return LspError.HeaderFieldTooLong,
             else => return err,
         } orelse return error.PrematureEndOfStream;
+        try stream.skipUntilDelimiterOrEof('\n');
+
+        debug.warn("Read line from stdin: {}\n", .{line});
 
         if (line.len <= 1) { // \r
```
Comment (xackus) [inline src/protocol.zig]:
```suggestion
        if (line.len == 0) {
```

File: src/protocol.zig [hunk 2]
```diff
@@ -14,37 +14,47 @@ pub const LspError = error{
...
-    if (bytesRead != jsonLen.?) {
+    var bytes_read: usize = 0;
+
+    while (bytes_read < json_len.?) {
+        bytes_read += try stream.read(jsonStr[bytes_read..]);
+    }
+
+    if (bytes_read != json_len.?) {
         return LspError.PrematureEndOfStream;
     }
```
Comment (xackus) [inline src/protocol.zig]: instead of writing a loop, use `readAll`

**Review by xackus — CHANGES_REQUESTED** [review]
Thanks for the PR!

You are right about the hack. I have already submitted a PR to upstream.

I think the warns will be too verbose.
Please remove them (except for the one that prints on errors).
If you are using vscode you can turn on logging. https://code.visualstudio.com/api/language-extensions/language-server-extension-guide#logging-support-for-language-server

**Comment by Sergeeeek** [comment]
Thanks for the feedback, fixed all of your comments

**Pull request merged**